src/NLog/LogFactory.cs

```diff
@@ -90,14 +90,53 @@ public class LogFactory : IDisposable
         /// </remarks>
         public event EventHandler<LoggingConfigurationChangedEventArgs> ConfigurationChanged;
 
-        private static event EventHandler<EventArgs> LoggerShutdown;
+        private static event EventHandler<EventArgs> _loggerShutdown;
 
         /// <summary>
-        /// Initializes static members of the LogManager class.
+        /// Event that is raised when the logging system is shutting down.
         /// </summary>
-        static LogFactory()
+        public static event EventHandler<EventArgs> LoggerShutdown
         {
-            RegisterEvents(DefaultAppEnvironment);
+            add
+            {
+                if (_loggerShutdown == null && DefaultAppEnvironment != null)
+                {
+                    try
+                    {
+                        DefaultAppEnvironment.ProcessExit += OnLoggerShutdown;
+                        InternalLogger.Debug("Registered shutdown event handler for ProcessExit.");
```

Comment:
Can `InternalLogger.Debug` be moved up so it happens before the subscription to ProcessExit ?